Custom repositioning of popup at edge

bijancamp opened the pull request:
Hi Per! I have added support for custom repositioning of the menu or button whenever they are outside of the window. This helps to prevent the menu from covering the selection, or the auto button from appearing directly under the mouse, whenever the selection is next to one of the window’s edges, as per our earlier discussion in Issue #15. My solution looks a little messy in "positioning.js" with nested "if" statements, so I would be very eager to make these changes cleaner if possible.

I chose to add these features as style config options in order to keep the number of main options in "options.js" at a minimum. Also, I thought it would be nice to have these settings in the CSS style, since that is where custom positioning is already done by the user.

For example, to shift the menu 55 pixels down whenever it goes outside on the top, you can add ```"menu_edge_top":"55px"``` to ```/*CONFIG_START{ ... }CONFIG_END*/```. If you want the current behavior of automatically repositioning to fit inside the window, then you can use ```"menu_edge_top":"auto"```. When set to a numerical value such as "55px", the element is shifted without any automatic repositioning done beforehand.

As an example, I can add the following to the "Icons only" styling preset to reposition the menu so that it does not cover the selection whenever the selection is near the top of the window:
```
.mainmenu {
  margin-top: -8px;
  margin-left: -45px;
}
/*CONFIG_START{"submenu_position":"topright",
               "submenu_corner":"bottomleft",
               "menu_edge_right":"auto",
               "menu_edge_bottom":"auto",
               "menu_edge_left":"auto",
               "menu_edge_top":"55px"}CONFIG_END*/
```
Here is what this looks like:

![selection search custom repositioning 1 nov 07 2015](https://cloud.githubusercontent.com/assets/8800015/11014161/a119c2bc-84f0-11e5-93d1-7986d603c2d3.gif)

I can also also reposition the auto button so that it does not appear under the mouse whenever the selection is near the bottom of the window:
```
.mainmenu > li:first-child{
 display: none;
}
.button{
  margin-top: 36px;
  margin-left: -24px;
}
.mainmenu {
  margin-top: -8px;
  margin-left: -45px;
}
/*CONFIG_START{"submenu_position":"topright",
               "submenu_corner":"bottomleft",
               "button_edge_right":"auto",
               "button_edge_bottom":"-38px",
               "button_edge_left":"auto",
               "button_edge_top":"auto"}CONFIG_END*/
```

And here is a demo of this:

![selection search custom repositioning 2 nov 07 2015](https://cloud.githubusercontent.com/assets/8800015/11014178/42b455f6-84f1-11e5-8cf9-f162d28509a6.gif)

Pitmairen commented:
Hi,
I think this looks quite good and it seems to work well.  The style config is probably the correct place to have these settings as they are somewhat dependent on the style in use.

But I have a few suggestions:

In popup.js:
There is no need to pass the style object as an argument to the _showPopupNode function. 
```js
_showPopupNode(x, y, node, style)
```
This function is defined as a inner function inside Popup so every variable that is declared in Popup's "main" scope will be directly accessible to all the inner functions. You can see this in _showPopupNodeRelativeToAnchor which also uses the style object.


In style.js:
You can reduce some of the repetition by using a loop. Something like this:
```js
var properties = ['menu_edge_right', 'menu_edge_right', 'menu_edge_left', ...];

for(var i in poperties){
    var prop = properties[i];
     if(config.hasOwnProperty(prop) &&
            config[prop].match(/^ *[+-]?((\d+(\.\d*)?)|(\.\d+)) *(px)? *$/)){
            _styleConfig[prop] = config[prop];
        }
}
```
The regexp can also probably be simplified a bit by maybe only searching for 
integer values. Float values are probably not needed, and can easily be added later if necessary. 
It can also be done without regexp with something like this:
```js
if(config.hasOwnProperty(prop) && 
  !isNaN(parseInt(config[prop])) && config[prop].endsWith("px"){
}
```

In positioning.js:
I don't think it's that much that can be done to clean up the ifs, but one way is to 
create a new private function that does some of the work. Something like this:

```js
function _calculateDiff(posValue, boundsValue, configValue){
    if (configValue === "auto")
        return boundsValue - posValue;
    else
        return parseInt(configValue);
}

// Then this can be used inside checkPosition like this:
if (pos.bottom > bounds.bottom) { // Goes outside on the bottom
    diffY = _calculateDiff(pos.bottom, bounds.bottom, diffConfig.bottom)
} else if (pos.top < bounds.top) { // Goes outside on the top
    diffY = _calculateDiff(pos.top, bounds.top, diffConfig.top)
}
.
.
.
```
The same with the creation of the  diffConfig variable. This can also be split into a 
new private function that creates the config so inside checkPosition it becomes:
var diffConfig = _getDiffConfig(style, node);

 
Otherwise I think it looks good, and I think this is a nice feature. :)

bijancamp commented:
Thank you for your suggestions! I will be working today on incorporating your changes. :D

bijancamp commented:
I have updated the changes according to your suggestions. I used regex instead of ```parseInt``` and ```.endsWith``` in order to allow both plain numbers and numbers with "px" on the end as valid values. I wasn't sure how to do this without regex.

I also added some explanation on using the style config in the customize section of the Options page. Since the explanation is kinda long, I made the page automatically scroll to the beginning of the customizing section whenever the "Customize" button is clicked, so that the explanation box does not keep the CSS textarea from being out of focus.

bijancamp commented:
I just made some minor revisions to the explanation for how to use the new feature.

Pitmairen commented:
I think looks good now, so i'm merging it. I will release a new version with the latest changes soon.

Thanks again for your excellent contribution! :D

bijancamp commented:
Thank you! I'm looking forward to making more contributions to this excellent extension! :D
